lib/phobos/actions/process_message.rb

```diff
@@ -64,6 +62,21 @@ def process_message(payload)
           end
         end
       end
+
+      def wait_for_next_retry(interval)
+        MAX_SLEEP_INTERVAL = 3
+        raise Phobos::AbortError if @listener.should_stop?
+        @listener.send_heartbeat_if_necessary
+        while interval > MAX_SLEEP_INTERVAL
+          sleep MAX_SLEEP_INTERVAL
+          raise Phobos::AbortError if @listener.should_stop?
+          @listener.send_heartbeat_if_necessary
+          interval = interval - MAX_SLEEP_INTERVAL
+        end
+        sleep interval
+        raise Phobos::AbortError if @listener.should_stop?
+        @listener.send_heartbeat_if_necessary
+      end
```

Comment:
Could be simplified to:

```ruby
def wait_for_next_retry(interval)
  while interval > 0
    raise Phobos::AbortError if @listener.should_stop?
    @listener.send_heartbeat_if_necessary
    sleep [interval, MAX_SLEEP_INTERVAL].min
    interval -= MAX_SLEEP_INTERVAL
  end
  raise Phobos::AbortError if @listener.should_stop?
  @listener.send_heartbeat_if_necessary
end
```

If you're comfortable with the following change to `Listener`, the code becomes even more readable IMO:
```ruby
# In `Listener`
def send_heartbeat_if_necessary
  raise Phobos::AbortError if should_stop?
  @consumer&.send_heartbeat_if_necessary
end

# In `ProcessMessage`
def wait_for_next_retry(interval)
  while interval > 0
    @listener.send_heartbeat_if_necessary
    sleep [interval, MAX_SLEEP_INTERVAL].min
    interval -= MAX_SLEEP_INTERVAL
  end
  @listener.send_heartbeat_if_necessary
end